### ISSUE TITLE
Requests will not completed, if all Logging Options are enabled

### ISSUE THREAD

**Issue by lexon0011** [IdentityServer/IdentityServer3#2481]
If we enable all logging options, the IdentityServer3 does not response after some requests.

```cs
new LoggingOptions()
	{
		EnableHttpLogging = true,
		EnableKatanaLogging = true,
		EnableWebApiDiagnostics = true,
		WebApiDiagnosticsIsVerbose = true
	};
``` 
The problem could be reproduce with the following steps:

--> Set all LoggingOptions to true (see above)
--> Browse to https://<url>/identity/.well-known/openid-configuration 
--> The configuration is shown as JSON in the browser
--> Now press very quickly and often the F5, to produce many requests
--> After some requests, the server no longer responds
--> If you change all LoggingOptions to false, all requests will be processed

Is it possible that the logging causes deadlocks?

We use
--> IdentityServer 2.3.0
--> .NET Framework 4.6

Thanks for helping.

**Comment by brockallen** [comment]
Do you have any data in the logs that might shed some light on the issue?

**Comment by lexon0011** [comment]
In the attached file you can see the log output: [Logfile.txt](https://github.com/IdentityServer/IdentityServer3/files/104999/Logfile.txt)

I could not find any error messages, which helps to solve the problem. Could you reproduce the problem on your environment?

**Comment by brockallen** [comment]
Nope, I couldn't see anything. What are you using for logging? Perhaps that's the issue.

**Comment by brockallen** [comment]
If you have further issues/questions related to this, feel free to reopen.

**Comment by MikaelGRA** [comment]
Hello @brockallen

I've experienced something frighteningly similar to this, and am also able to very consistently reproduce it.

My issue also ONLY occurrs when the logging options are enabled, and I am using Trace for logging in my application (I am using Trace through NLog).

Basically, if I try to request 5 refresh token requests at the same time, based on the same refresh token (because I don't know any better!), then the application becomes deadlocked (atleast the OAuth2 part of it).

Here's some code that, pretty much without fail, reproduces the issue for me locally (I was unable to reproduce it with @lexon0011 procedure):

```C#

      static async Task PerformStressTest()
      {
         var client = new TokenClient(
            endpoint + "/oauth/connect/token",
            "client-app-id",
            "client-app-secret" );

         var response = await client.RequestResourceOwnerPasswordAsync( "username", "password", "openid profile email my_scope offline_access" );
         string refreshToken = response.RefreshToken;

         while( true )
         {
            List<Task<TokenResponse>> tasks = new List<Task<TokenResponse>>();
            for( int i = 0 ; i < 5 ; i++ )
            {
               tasks.Add( client.RequestRefreshTokenAsync( refreshToken, null ) );
            }

            try
            {
               await Task.WhenAll( tasks );
            }
            catch( Exception )
            {
               // really dont care
            }

            var completedTask = tasks.First( x => x.IsCompleted && !x.Result.IsError );

            refreshToken = completedTask.Result.RefreshToken;

            await Task.Delay( 1000 );
         }
      }
```

The odd thing, is that it seems to block other parts of my code that uses logging as well.

// Mikael Guldborg

**Comment by lexon0011** [comment]
I use Log4Net and the problem still exists in my project.

**Comment by brockallen** [comment]
If you can put that into a unit/integration test to repro it, that'd be a help in getting this sorted.

**Comment by MikaelGRA** [comment]
Hello @brockallen 

I have created a repository that reproduces the error here:

https://github.com/MikaelGRA/IdentityServer3.Deadlocked

Hopefully, you can use this.

It reminds a tiny bit of my own project setup.

You will need to call Update-Database (to create a LocalDB for aspnet identity/entity framework)

Once you have done that you can start the "Deadlocked" project, followed by the "Deadlocked.Client" project. You will find that you never hit the:

```C#
await Task.Delay( 1000 );
```

In the Deadlocked.Client project. (at some point a timeout exception is thrown).

After this deadlocked has been hit, no OAuth2 operations on the server works until you restart the project.

What you will probably find interesting is that if you do ANY of the following things, the deadlock will NOT occur:
 * Comment out the "target" and "logger" in the NLog.config file.
 * Comment out the LoggingOptions in the Startup.Auth, as suggested by @lexon0011 
 * Change the loop in the client application, so it only iterates once, basically eliminating all concurrency.

Let me know if you have any issues getting it running.

**Comment by leastprivilege** [comment]
What happens if you use a different logger - e.g. Log4Net?

Same deadlock?

**Comment by MikaelGRA** [comment]
Hey,

I have tried adding log4net into a new branch in the repro project:

https://github.com/MikaelGRA/IdentityServer3.Deadlocked/tree/log4net

It seems that if I configure log4net up to use Trace, then the same issue occurs. If I dont use Trace, then it works just fine. This is exactly the same behaviour as NLog.

Disclaimer: I have never used log4net before, so I am defintely not certain I am using it correctly.

**Comment by totpero** [comment]
Hey,

I have the something like this but in other context, with external providers.
 I enable:
 `LoggingOptions = new LoggingOptions {EnableKatanaLogging  = true, EnableHttpLogging = true}`
and i have problem with yahoo external provider (Caption text empty) and Twitter certificate error, and with this logging enable if i press one of this two external provider button, my idsrv3 site loading until i press stop button from browser.
If i disable this logging options, i get the correct error.
With this options enable the site infinite loading and no error show.
I hope to help my testcase in this issue.

**Comment by brockallen** [comment]
I downloaded the repro from @MikaelGRA and I could not reproduce the deadlock. I see that the client code is trying to use a refresh token multiple times in parallel. Is that the scenario that you're trying to model, @MikaelGRA?

If so, then issue has come up before: https://github.com/IdentityServer/IdentityServer3.EntityFramework/issues/42. The conclusion from the other thread was that your client should not be trying to use the same refresh token in parallel.

As for the OP @lexon0011, do you have anything that can repro this issue?

**Comment by MikaelGRA** [comment]
Hey @brockallen 

It is indeed the scenario I am trying to model, as it is one of the factors that seem to be causing the **deadlock**.

I agree that the client **SHOULD NOT** be doing it. However, I, as the OAuth2 server, have absolutely no way to prevent clients from doing it (I am not the client). And the fact that it produces a deadlock on the server, that shuts down all subsequent requests until the server is restarted seems to be a **very** serious issue, that is not at all being referred to in the issue you are referencing.

In fact, I already read about that issue before I made my comment here, and I completely agree with the fact that if you make multiple concurrent refresh token requests, only 1 of them should have any chance of succeeding.

Now, as to why you cannot reproduce it. This seems odd to me. I will see if I can reproduce the issue on some of my colleagues computers tomorrow.

But my initial thoughts would be that, maybe if you try to increase the concurrency, the chances would be higher?

I think that with concurrency, it is difficult to produce a determinstic repro. However, every single time I have tried with 5 concurrent requests, the server has deadlocked.

**Comment by brockallen** [comment]
> But my initial thoughts would be that, maybe if you try to increase the concurrency, the chances would be higher?

I had already bumped up the client to 15 concurrent tasks. :/

**Comment by brockallen** [comment]
I'd suggest: 

@lexon0011 reopen if you can reproduce the error situation. something that we can run to reproduce it would help greatly.

@MikaelGRA open a new issue if you feel strongly about the scenario where a client tries to use a refresh token multiple times in parallel since that's a slightly different issue.

**Comment by MikaelGRA** [comment]
**Regarding my comment yesterday**: I was able to reproduce the error on my colleagues computers, so I am very confused as to why it will not be reproduced for you.

I did find that there were some issues in my repro project with two files that were not checked in. For that I apologize. I have checked these files in now.

@brockallen: I do not know if I agree that this is a different issue than the one the OP originally suggested, since this is an issue that in the end, revolves around concurrency and the log implemented in IdentityServer3. The  fact that I am able to reproduce it by sending concurrent refresh token requests is merely a side effect. I am also able to reproduce it if I send concurrent access token requests (with username/password), while all the logging features are enabled. This is probably the case for a lot of other scenarios as well.

That is why I am not going to be opening a new issue. I do feel that this is something you should investigate, given the severity. As it is right now though, it is not much of an issue for me, since I can simply disable these logging features in the production environment (something I have already done).

I do have a nagging worry in my head though, that maybe disabling these logging features are simply *reducing the chances* of a deadlock occurring, since this seem to disable the logging of an extraordinary amount of things. :/

**Comment by brockallen** [comment]
> I did find that there were some issues in my repro project with two files that were not checked in. For that I apologize. I have checked these files in now.

Yep, I worked around those :)

>  I do not know if I agree that this is a different issue than the one the OP originally suggested, since this is an issue that in the end, revolves around concurrency and the log implemented in IdentityServer3

You might be right, but at the same time I do think there's another issue. So I think it's 2 potentially related issues. We'll see as we move forward on this...

**Comment by robdmoore** [comment]
I have an identity server 3 instance that has WSFederation configured as an external identity provider. I just found that if I set `LoggingOptions.EnableHttpLogging` to `true` then the call to `/external?provider=adfs&signin={signinid}` hangs and never returns.

**Comment by arindamdat** [comment]
@lexon0011 :+1: You just saved me!!!! I still wonder how did you arrive into the conclusion that those LoggingOptions were causing the deadlock...
@brockallen : Eagerly waiting to know the root cause and it's fix!!
Cheers!!!

**Comment by brockallen** [comment]
> I have an identity server 3 instance that has WSFederation configured as an external identity provider. I just found that if I set LoggingOptions.EnableHttpLogging to true then the request to /external?provider=adfs&signin={signinid} hangs and never returns, but the other requests before it worked.

Sorry this was missed because this issue was closed. If you're still having an issue, can you open a new issue? Thx
